File: engine/src/main/java/de/hanno/hpengine/util/commandqueue/FutureCallable.java
package de.hanno.hpengine.util.commandqueue;

import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;

public abstract class FutureCallable<RETURN_TYPE> extends CompletableFuture<Callable<RETURN_TYPE>> {

    public final CompletableFuture<RETURN_TYPE> getFuture() {
        return (CompletableFuture<RETURN_TYPE>) this;
    }

    public abstract RETURN_TYPE execute() throws Exception;

}

File: engine/src/main/kotlin/de/hanno/hpengine/util/commandqueue/FutureCallable.kt
package de.hanno.hpengine.util.commandqueue

import java.lang.Exception
import java.util.concurrent.Callable
import java.util.concurrent.CompletableFuture

abstract class FutureCallable<RETURN_TYPE> : CompletableFuture<Callable<RETURN_TYPE>?>() {
    val future: CompletableFuture<RETURN_TYPE>
        get() = this as CompletableFuture<RETURN_TYPE>

    @Throws(Exception::class)
    abstract fun execute(): RETURN_TYPE
}